Feature/wg 109

Evil00787 opened the pull request:
Allow me, stasgora!
Adding redo and undo!

Evil00787 committed:
Added JumpCommand

Evil00787 committed:
Fixed issues with GRAPH controllers - HistoryController

Evil00787 committed:
Fixing keyboard

Evil00787 committed:
Modified Processor

Evil00787 committed:
History now works in alpha version! Try it now! (Q - undo, E - redo)

stasgora commented on Assets/Scripts/Controllers/HistoryController.cs:
```diff
@@ -0,0 +1,58 @@
+using Services.History;
+
+namespace Controllers
+{
```
- Configure your IDE code formatting to be consistent with the rest of the code
- Reformat all added files

stasgora commented on Assets/Scripts/Controllers/HistoryController.cs:
```diff
@@ -0,0 +1,58 @@
+    public class HistoryController: MonoBehaviour
+    {
+        private Stack<ICommand> _Undocommands = new Stack<ICommand>();
+        private Stack<ICommand> _Redocommands = new Stack<ICommand>();
```
- Inconsistent naming
- Use readonly whenever you can
- Use more meaningful names
```suggestion
        private readonly Stack<ICommand> redoActionStack = new Stack<ICommand>();
```

stasgora commented on Assets/Scripts/Controllers/HistoryController.cs:
```diff
@@ -0,0 +1,58 @@
+            nodeController = GetComponent<NodeController>();
+        }
+
+        public void Redo(int levels)
```
I don't think the `levels` parameter is necessary since user triggers an action by pressing a button - on at a time

stasgora commented on Assets/Scripts/Controllers/HistoryController.cs:
```diff
@@ -0,0 +1,58 @@
+
+        #region UndoHelperFunctions
+
+        public void InsertInUnDoRedoForJump(Model.Node prevNode, Model.Node targetNode)
```
Generify:
- Use more meaningful name like `RegisterAction`
- Pass a created action `ICommand` argument (this function should be used to register any action)
```suggestion
        public void RegisterAction(ICommand action)
```

stasgora commented on Assets/Scripts/Controllers/HistoryController.cs:
```diff
@@ -0,0 +1,58 @@
+        public void InsertInUnDoRedoForJump(Model.Node prevNode, Model.Node targetNode)
+        {
+            ICommand cmd = new JumpCommand(prevNode, targetNode, ref nodeController);
+            _Undocommands.Push(cmd); _Redocommands.Clear();
```
**Formatting!**

stasgora commented on Assets/Scripts/Controllers/HistoryController.cs:
```diff
@@ -0,0 +1,58 @@
+
+namespace Controllers
+{
+    public class HistoryController: MonoBehaviour
```
- Configure your indents to tabs of size 4 - like the rest of code
- Reindent all changed lines

stasgora replied:
If this was Python, you wouldn't get away with this :P

stasgora commented on Assets/Scripts/Controllers/NodeController.cs:
```diff
@@ -107,7 +107,7 @@ public class NodeController: MonoBehaviour {
 		}
 
-		private void SetNodeState(Node node, NodeState state) {
+		public void SetNodeState(Node node, NodeState state) {
```
This is an internal function and therefore must stay private. Maybe a public `ForceSetSelectedNode` function would be a good idea - it will come in handy once we will be implementing a search functionality. We can set a node as active and then select it there.

stasgora commented on Assets/Scripts/Controllers/InputController.cs:
```diff
@@ -19,11 +19,13 @@ public class InputController : MonoBehaviour {
 		public NodeController NodeController { get; private set; }
 		public CameraController CameraController { get; private set; }
 		public GraphController GraphController { get; private set; }
+        public HistoryController HistoryController { get; private set; }
```
**Indents!**

stasgora commented on Assets/Scripts/InputModule/Binding/PcInputBinding.cs:
```diff
@@ -19,5 +19,8 @@ public class PcInputBinding: InputBinding {
 		public PCButton InfographicToggle;
 		public PCButton ConnectionModeToggle;
+
+        public PCButton Redo;
```
Naming; sth like `UndoUserAction`

stasgora commented on Assets/Scripts/InputModule/Processor/InputProcessor.cs:
```diff
@@ -8,21 +8,28 @@ public abstract class InputProcessor {
 		protected void ExitNodeTraverseMode() => Controller.GraphController.GraphMode.Value = GraphMode.FREE_FLIGHT;
 
-		protected void OnNodeChosen(Ray ray) {
+        protected void Redo() { Debug.Log("Button pressed Redo!"); Controller.HistoryController.Redo(1); }
```
- Formatting
- Naming - like button names
- We have our own logger: `Services.Logger`; see some usages
- Remove all spamming debug logs before merge

stasgora commented on Assets/Scripts/InputModule/Processor/InputProcessor.cs:
```diff
@@ -8,21 +8,28 @@ public abstract class InputProcessor {
-				Controller.NodeController.SelectedNode = GraphController.Graph.GetNodeFromObject(raycastHit.collider.gameObject);
-			}
+            if (RaycastNode(ray, out raycastHit)) {
+                Controller.HistoryController.InsertInUnDoRedoForJump(Controller.NodeController.SelectedNode, GraphController.Graph.GetNodeFromObject(raycastHit.collider.gameObject));
```
At this point we are in Input module, a selected node can be:
- Invalid (Disabled)
- Already selected (When switching connection mode)

Instead create a `JumpCommand` after the `NodeController` validation by subscribing to the `SelectedNode` changes (`OnSelectedNodeChanged`)

stasgora commented on Assets/Scripts/InputModule/Processor/InputProcessor.cs:
```diff
@@ -31,5 +38,7 @@ public abstract class InputProcessor {
 		
 		private bool RaycastNode(Ray ray, out RaycastHit hit) => Physics.Raycast(ray, out hit, float.MaxValue, LayerMask.GetMask("Node"));
+
+
```
Formatting

stasgora commented on Assets/Scripts/Services/History/CommandInterface.cs:
```diff
@@ -0,0 +1,8 @@
+﻿namespace Services.History
+{
+    public interface ICommand
```
That name tells nothing; `UserAction` would be better

stasgora commented on Assets/Scripts/Services/History/JumpCommand.cs:
```diff
@@ -0,0 +1,50 @@
+{
+    public class JumpCommand : ICommand
+    {
+        private Model.Node _prevNode;
```
- Formatting
- Consistent naming: e.g. old/new selected
- Use normal imports

stasgora commented on Assets/Scripts/Services/History/JumpCommand.cs:
```diff
@@ -0,0 +1,50 @@
+
+namespace Services.History
+{
+    public class JumpCommand : ICommand
```
More like `NodeSelectedAction`

stasgora commented on Assets/Scripts/Services/History/JumpCommand.cs:
```diff
@@ -0,0 +1,50 @@
+    {
+        private Model.Node _prevNode;
+        private Model.Node _targetNode;
+        public NodeController nodeController;
```
This is a service layer, it shouldn't know nor care about what program or environment it is running in. Instead create an Action or delegate with `Node` parameter that will be setting the `SelectedNode` (in fact it would be binded to the force function I suggested earlier) It may be even static since it will be the same for every action instance

stasgora commented on Assets/Scripts/Services/History/JumpCommand.cs:
```diff
@@ -0,0 +1,50 @@
+
+        public void Execute() //do action forward
+        {
+            nodeController.SetNodeState(_targetNode, Model.NodeState.ACTIVE);
```
Always extract duplicated functionality to a private utility function

stasgora commented on Assets/Scripts/Services/History/JumpCommand.cs:
```diff
@@ -0,0 +1,50 @@
+            _prevNode = prevNode;
+            _targetNode = targetNode;
+            nodeController = controller;
+          // if(_prevNode == null) Debug.Log("Created history: null -> "+ _targetNode.ID+ " | "+ _targetNode.Title);
```
Do not leave commented out code

stasgora commented on Assets/Scripts/Controllers/HistoryController.cs:
```diff
@@ -0,0 +1,58 @@
+{
+    public class HistoryController: MonoBehaviour
+    {
+        private Stack<ICommand> _Undocommands = new Stack<ICommand>();
```
Separate unity specific stuff (Controller based) from the business logic (Services).

##### First a little backstory about code modularization & architecture: 
- A user action history is a service that records actions and allows to undo & redo them (and so those would be its public methods). 
- A history controller is a client that uses this service functionality by implementing user-action interface to fit clients needs (e.g. undoing user node selections - note unity context).

So in this case a `HistoryController` tasks are to:
- Subscribe to `SelectedNode` changes
- Instantiate implementation of user-action interface
- Register it in service

On top of that it needs to inform action implementation how it should handle it's execution - so it binds its static delegate/Action to the appropriate function in `NodeController` (as mentioned previously). This way the service layer stays independent & clueless about what its doing.

stasgora reviewed (changes requested):
Great job, just needs some work ;)

stasgora commented on Assets/Scripts/InputModule/Processor/PCInputProcessor.cs:
```diff
@@ -18,6 +18,9 @@ public class PCInputProcessor: InputProcessor {
 			binding.InfographicToggle.OnPress += () => Controller.GraphController.Infographic.SetActive(!Controller.GraphController.Infographic.activeSelf);
 			binding.ConnectionModeToggle.OnPress += () => Controller.GraphController.SwitchConnectionMode();
+
+            binding.Undo.OnPress += Undo;
```
- Add Cave buttons with the same bindings
- Assign them according to our specification

Evil00787 committed:
Resolved all problems with the architecture.
Created HistoryService.
Buisness logic and Unity components are finally separated.
